src/System.IO.Pipelines/src/System/IO/Pipelines/Pipe.cs

```diff
@@ -463,17 +457,22 @@ private void AdvanceReader(BufferSegment consumedSegment, int consumedIndex, Buf
                     _readerAwaitable.Reset();
                 }
 
-                while (returnStart != null && returnStart != returnEnd)
-                {
-                    returnStart.ResetMemory();
-                    ReturnSegmentUnsynchronized(returnStart);
-                    returnStart = returnStart.NextSegment;
-                }
-
                 _operationState.EndRead();
             }
 
+            // Scehudle early, prior to clean up
             TrySchedule(_writerScheduler, completionData);
+
+            // Reset the segments outside lock
+            BufferSegment returnLast = null;
+            while (returnStart != null && returnStart != returnEnd)
+            {
+                returnLast = returnStart;
+                returnStart = returnStart.NextSegment;
+                returnLast.ResetMemory();
+            }
+
+            ReturnSegment(returnLast);
```

Comment:
Would returning the segment so late after `TrySchedule` that it won't be able to ever pick it up in time would make polling effectively useless?

Reply:
Also moving to 1 segment causes questions around `DefaultMinimumSegmentSize` which is based on number of pooled segments; so will skip it in this PR

Reply:
Probably can still do something a bit better here though